### ISSUE TITLE
Custom log message during training

### ISSUE THREAD

**Issue by emjotde** [marian-nmt/marian-dev#195]
Just a random thought. Would be nice to able to modify the the log message during training

```
Ep. 1 : Up. 6000 : Sen. 1193481 : Cost 2.68 : Time 86.83s : 32988.54 words/s
```

By providing a template, e.g.

```
--log-template "Ep. {%EPOCH%} : Up. {%UPDATE%} : Sen. {%SAMPLES%} : Cost {%COST%} : Time {%BATCH_SECONDS%} : {%WORDS_PER_SECOND%} words/s
```

And them provide more options like:
```
{%LR%}
{%TOTAL_WORDS%}
{%TOTAL_TIME%}
...
```

Not too urgent by kind of cute.

**Comment by emjotde** [comment]
One could for instance generate latex tables from that directly :)

**Comment by emjotde** [comment]
@ugermann maybe?

**Comment by frankseide** [comment]
I locally implemented a modification that logs ce loss per target word, along with associated number of target words scored. This makes losses better comparable across systems and parallelization configurations, which I find useful. Example:
```
[2018-02-22 20:39:54] Ep. 1 : Up. 500 : Sen. 350559 : Cost 9.89 * 10696858 : Time 482.42s : 22173.46 words/s : L.r. 9.3750e-06
[2018-02-22 20:47:18] Ep. 1 : Up. 1000 : Sen. 699788 : Cost 8.54 * 10581251 : Time 444.68s : 23794.96 words/s : L.r. 1.8750e-05
[2018-02-22 20:54:42] Ep. 1 : Up. 1500 : Sen. 1047256 : Cost 7.92 * 10575207 : Time 443.88s : 23824.48 words/s : L.r. 2.8125e-05
```
This requires to recover the normalization done in `Cost()` and undo it. Here is the modified code in `scheduler.h`:
```
  void update(float cost, Ptr<data::Batch> batch) {
    auto batchSize = batch->size();
    auto batchTargetWords = batch->words(-1);
    // reconstruct sum cost
    auto costType = options_->get<std::string>("cost-type");
    cost *= // what was cost normalized with?
      /*if*/ (costType == "ce-sum") ?
        1
      /*else if*/ : ((costType == "ce-mean-words") ?
        batchTargetWords
      /*else*/ :  // use ce-mean for all others (not correct for some)
        batchSize);
    // undo averaging over workers
    cost *= options_->get<std::vector<size_t>>("devices").size();

    costSum += cost;               // aggregate cost since last display
    wordsDisp += batchTargetWords; // targets processed since last display
    samples += batch->size();      // sentences processed in this epoch
    samplesDisp += batch->size();  // (unused)
    state_->newBatch();

    if(state_->batches % options_->get<size_t>("disp-freq") == 0) {
      if(options_->get<bool>("lr-report")) {
        LOG(info,
            "Ep. {} : Up. {} : Sen. {} : Cost {:.2f} * {} : Time {} : {:.2f} "
            "words/s : L.r. {:.4e}",
            state_->epochs,
            state_->batches,
            samples,
            costSum / wordsDisp, wordsDisp, // cost per target word
            timer.format(2, "%ws"),
            wordsDisp / std::stof(timer.format(5, "%w")),
            state_->eta);
      } else {
        LOG(info,
            "Ep. {} : Up. {} : Sen. {} : Cost {:.2f} * {} : Time {} : {:.2f} "
            "words/s",
            state_->epochs,
            state_->batches,
            samples,
            costSum / wordsDisp, wordsDisp,
            timer.format(2, "%ws"),
            wordsDisp / std::stof(timer.format(5, "%w")));
      }
      timer.start();
      costSum = 0;
      wordsDisp = 0;
      samplesDisp = 0;
    }

    validated_ = false;
  }
```
It also requires a minor change to `CorpusBatch::words()` to accept a parameter to select the stream, where `-1` stands for the last stream = target.
```
size_t words(int which = 0) const { return batches_[which >= 0 ? which : which + (ptrdiff_t)batches_.size()]->batchWords(); }
```

**Comment by kpu** [comment]
I too have wanted more statistics (hi @fiqas) which could be computed in a more natural way by adding them to the computation graph.  Which runs into the single root problem.  

Though admittedly this could be solved by extracting the cost before it goes into normalization.

**Comment by emjotde** [comment]
@kpu we have a training process observer that records statistics about the process, just need to add more fields.

**Comment by kpu** [comment]
So if I wanted weighted perplexity and unweighted perplexity logged, how would I go about doing that using the same implementation of the Cost function?

**Comment by fiqas** [comment]
Well, if you try to log two costs at the same time, you cannot do it because it will create two roots.
Currently, I print it like:
1) cost = calculate_cost()
2) print it
3) cost = cost * 0 + calculate_cost()

Which works, but is awkward and band-aid-ish.

**Comment by emjotde** [comment]
That's not something you will get around, that's just the way the algorithm works.

**Comment by emjotde** [comment]
@fiqas @kpu Ha, just realized you could probably do that by making one part of the graph inference-only.

```c++
auto g = New<ExpressionGraph>()
auto logits = build_net(g, ...); 

g->setInference(true);
auto dummyCost = calc_report_cost(g, logits, ...);
g->setInference(false);

auto cost = calc_cost(g, logits, ...);

g->forward();

float cost1 = dummyCost->scalar();
float cost2 = cost->scalar();

g->backward();

# do stuff with cost1 and cost2
```

That should actually work and not trigger the multiple node for backward computation exception. Forgot about that. Just make sure to surround the cost you do not want to

**Comment by fiqas** [comment]
Oh, that will be pretty useful! This way, we can calculate additional stuff just for logging.

**Comment by emjotde** [comment]
Just be careful to not cut off the root form the rest of the graph, for instance by giving the cost only non-trainable nodes as inputs, as that would stop back-propagation. Also remember that a node is non-trainable if ALL it's inputs are non-trainable, so it propagates.

**Comment by emjotde** [comment]
I am not quite sure if it fully works in master. If not let me know and a I have version where it is tested. There I am using this to switch between different parts of the model for different mini-batches and for experiments with teacher-student models, it's quite powerful.

**Comment by ugermann** [comment]
Going back to the original issue, it looks reasonably straightforward. I'll look into it next week (currently travelling).
